google/cloud/storage/client.cc

```diff
@@ -50,58 +48,97 @@ This is often a problem because:
     msg += file_name;
     google::cloud::internal::RaiseRuntimeError(msg);
   }
-  if (not request.HasOption<DisableMD5Hash>() and
-      not request.HasOption<MD5HashValue>()) {
-    // Open a separate stream to read the file, because once we hit EOF there
-    // is no guarantee we can rewind the file to the beginning.
-    std::ifstream is2(file_name);
-    MD5_CTX md5;
-    MD5_Init(&md5);
-    std::string buffer;
-    buffer.resize(raw_client_->client_options().upload_buffer_size(), '\0');
-    while (not is2.eof()) {
-      is2.read(&buffer[0], buffer.size());
-      MD5_Update(&md5, &buffer[0], static_cast<std::size_t>(is2.gcount()));
-    }
 
-    std::string hash(MD5_DIGEST_LENGTH, ' ');
-    MD5_Final(reinterpret_cast<unsigned char*>(&hash[0]), &md5);
-    request.set_option(
-        MD5HashValue(internal::OpenSslUtils::Base64Encode(hash)));
+  std::string payload(std::istreambuf_iterator<char>{is}, {});
+  request.set_contents(std::move(payload));
+
+  return raw_client_->InsertObjectMedia(request).second;
+}
+
+ObjectMetadata Client::UploadFileResumable(
+    std::string const& file_name,
+    google::cloud::storage::internal::ResumableUploadRequest const& request) {
+  std::ifstream source(file_name);
+  if (not source.is_open()) {
+    std::string msg = __func__;
+    msg += ": cannot open source file ";
+    msg += file_name;
+    google::cloud::internal::RaiseRuntimeError(msg);
   }
+  // This function only works for regular files, and the `storage::Client()`
+  // class checks before calling it.
+  std::uint64_t source_size = google::cloud::internal::file_size(file_name);
 
-  auto result = raw_client_->WriteObject(request);
-  auto streambuf = std::move(result.second);
+  auto result = UploadStreamResumable(source, source_size, request);
+  if (not result.first.ok()) {
+    google::cloud::internal::RaiseRuntimeError(result.first.error_message());
+  }
+  return result.second;
+}
 
-  std::string buffer;
-  buffer.resize(raw_client_->client_options().upload_buffer_size(), '\0');
-  while (not is.eof() and is.good()) {
-    is.read(&buffer[0], buffer.size());
-    streambuf->sputn(&buffer[0], is.gcount());
+std::pair<Status, ObjectMetadata> Client::UploadStreamResumable(
+    std::istream& source, std::uint64_t source_size,
+    internal::ResumableUploadRequest const& request) {
+  Status status;
+  std::unique_ptr<internal::ResumableUploadSession> session;
+  std::tie(status, session) = raw_client()->CreateResumableSession(request);
+  if (not status.ok()) {
+    return std::make_pair(status, ObjectMetadata{});
   }
-  auto response = streambuf->Close();
-  if (response.status_code >= 300) {
-    std::ostringstream os;
-    os << __func__ << ": error in during upload "
-       << Status(response.status_code, response.payload);
-    google::cloud::internal::RaiseRuntimeError(os.str());
+
+  // GCS requires chunks to be a multiple of 256KiB.
+  auto chunk_size = internal::UploadChunkRequest::RoundUpToQuantum(
```

Comment:
We can tune this in a future PR, but I want to note that having a chunk size too small will result in sending a silly amount of requests if the object is large.  We use a pretty large value in gsutil by default (100M):
https://github.com/GoogleCloudPlatform/gsutil/blob/88ab2023f0dcf3d7a5444832eb547b2cbc68d7bd/gslib/util.py#L833

Reply:
Ack.